sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,591 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Manages update artifacts stored on this sled. The implementation is a very
+//! basic content-addressed object store.
+//!
+//! GET operations are handled by the "Repo Depot" API, which is deliberately
+//! a separate Dropshot service from the rest of Sled Agent. This is to avoid a
+//! circular logical dependency, because we expect Sled Agent to fetch artifacts
+//! it does not have from another Repo Depot that does have them (at Nexus's
+//! direction). This API's implementation is also part of this module.
+//!
+//! POST, PUT, and DELETE operations are handled by the Sled Agent API.
+
+use std::io::ErrorKind;
+use std::net::SocketAddrV6;
+use std::sync::LazyLock;
+use std::time::Duration;
+
+use camino::{Utf8Path, Utf8PathBuf};
+use camino_tempfile::NamedUtf8TempFile;
+use dropshot::{
+    Body, ConfigDropshot, FreeformBody, HttpError, HttpResponseOk,
+    HttpServerStarter, Path, RequestContext, StreamingBody,
+};
+use futures::{Stream, TryStreamExt};
+use illumos_utils::zpool::ZPOOL_MOUNTPOINT_ROOT;
+use omicron_common::address::REPO_DEPOT_PORT;
+use omicron_common::disk::{DatasetKind, DatasetsConfig};
+use omicron_common::update::ArtifactHash;
+use repo_depot_api::*;
+use sha2::{Digest, Sha256};
+use sled_storage::manager::StorageHandle;
+use slog::{error, info, Logger};
+use tokio::fs::File;
+use tokio::io::AsyncWriteExt;
+
+const TEMP_SUBDIR: &str = "tmp";
+
+#[derive(Clone)]
```

Comment:
I think it'd be worth some comments explaining how this thing works.  Maybe something like:

```suggestion
/// Content-addressable local storage for software artifacts
// Storage for artifacts is backed by datasets that are explicitly designated for this purpose.  The `T: StorageBackend` parameter, which varies between the real and simulated sled agent, specifies exactly which datasets are available for artifact storage.  That's the only thing `T` is used for.  The behavior of storing artifacts as files under one or more paths is identical for all callers (i.e., both the real and simulated sled agents).
// A given artifact is generally stored on exactly one of the datasets designated for artifact storage.  There's no way to know from its key which dataset it's stored on.  This means:
// - for PUT, we pick a dataset and store it there
// - for GET, we look in every dataset for it until we find it
// - for DELETE, we attempt to delete it from every dataset (even after we've found it in one of them)
#[derive(Clone)]
```